### PR TITLE
Feat: discover nodes and number of GPUs in a cluster

### PR DESCRIPTION
**Pull request by bsraya**
Discover nodes and numbers of gpu on runtime.

Changes made in:
1. pkg/placement/placement_manager.go
2. pkg/scheduler/scheduler.go

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -20,24 +22,58 @@ import (
...
+	} else {
+		kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	}
+	flag.Parse()
+
+	// use the current context in kubeconfig
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		panic(err.Error())
+	}
```
Comment (heyfey) [inline pkg/scheduler/scheduler.go]: We don't need to build new config here, we can just pass `config *rest.Config` in `NewScheduler()` to it.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -20,24 +22,58 @@ import (
...
-	gpus = 8
-
+	databaseNameRunningJobs     = "runnings"
+	reschedChannelSize          = 100
+	restartChannelSize          = 100
 	rateLimitTimeMetricsSeconds = 5
 	reschedRateLimitSeconds     = 30
 )
 
+func DiscoverGPUs() int {
```
Comment (heyfey) [inline pkg/scheduler/scheduler.go]: I think we there's no need to expose this method, use `discoverGPUs` instead.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -20,24 +22,58 @@ import (
...
+
+	// create the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		panic(err.Error())
```
Comment (heyfey) [inline pkg/scheduler/scheduler.go]: It's better to return an error rather than `panic` here.

File: pkg/placement/placement_manager.go [hunk 4]
```diff
@@ -28,11 +33,39 @@ var launcherToleration = corev1.Toleration{
 	Effect:   corev1.TaintEffectNoExecute,
 }
 
-// nodes for testing
-// TODO
-var defaultNodes = map[nodeName]*nodeState{
-	"gpu3": NewNodeState("gpu3", 4),
-	"gpu4": NewNodeState("gpu4", 4),
+func DiscoverNodes() map[nodeName]*nodeState {
```
Comment (heyfey) [inline pkg/placement/placement_manager.go]:
```suggestion
func discoverNodes() map[nodeName]*nodeState {
```
Don't need to expose this method .

File: pkg/placement/placement_manager.go [hunk 5]
```diff
@@ -67,9 +100,11 @@ func NewPlacementManager(id string, kConfig *rest.Config) (*PlacementManager, er
 	podListerInformer := sharedInformers.Core().V1().Pods()
 	podInformer := podListerInformer.Informer()
 
+	defaultNodes := DiscoverNodes()
```
Comment (heyfey) [inline pkg/placement/placement_manager.go]:
```suggestion
	nodes := DiscoverNodes()
```

nodes are no longer hard-coded, so `defaultNodes` may not be a proper name.

File: pkg/placement/placement_manager.go [hunk 6]
```diff
@@ -28,11 +33,39 @@ var launcherToleration = corev1.Toleration{
...
+	nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+
+	for _, node := range nodes.Items {
+		gpus := node.Status.Capacity["nvidia.com/gpu"]
+		defaultNodes[nodeName(node.Name)] = NewNodeState(nodeName(node.Name), int(gpus.Value()))
+	}
+	return defaultNodes
```
Comment (heyfey) [inline pkg/placement/placement_manager.go]: Same as comment in `pkg/scheduler.go`
1. No need to build config
2. return error instead of `panic`

**Comment by heyfey** [comment]
Also, I think we need to modify ClusterRole in `deploy/vodascheduler/vodascheduler.yaml` to give vodascheduler the permission to `list nodes`

**Comment by bsraya** [comment]
Grant `vodascheduler` a permission to `list nodes`

**Comment by heyfey** [comment]
LGTM

Can you log about `GPUAvailable` in scheduler and `nodeStates` in placement manager?
You can add `klog.InfoS(...)` in `Run()` and have the log right after `"Starting scheduler", ...` or `"Starting placement manager", ...`

**Comment by bsraya** [comment]
Add `klog.InfoS(...)`

**Pull request merged**